### PR TITLE
VpnSiteLinkConnection NewAPIs for sharedkey

### PR DESCRIPTION
**Pull request by Khushboo-Baheti**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary> Click here to see the details of Step 1 </summary>

### Breaking changes review (Diagram Step 1)

If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
- The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
- You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

</details>

<details> 
<summary> Click here to see the details of Step 2 </summary>

### ARM API changes review (Diagram Step 2)

- For help with Step 2 / "ARM review", see https://aka.ms/azsdk/pr-arm-review.

</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [ ] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [x] Other, please clarify:
    - Adding sharedkey actions in new APIs of VpnsiteLinkConnection and new definitions for ConnectionSharedKey
ConnectionResetSharedKey return types of these actions.
    - Adding new property dpdtimeout in VpnsiteLinkConnection

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (visingla-ms) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2637]: >sharedKey [](http://example.com/codeflow?start=198&length=9)

The resourceType shuld be plural. Also, is sharedKey a resource type or an action name? I can see, you have defined both GET and POST on the same sharedKey name, which is fundamentally wrong.

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (visingla-ms) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2779]: >sharedKey/reset [](http://example.com/codeflow?start=198&length=15)

This is not right way to define a POST Action. Either it should be "/vpnLinkConnections/{linkConnectionName}/resetSharedKey" or "/vpnLinkConnections/{linkConnectionName}/sharedKeys/{sharedKeyName}/reset".

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (visingla-ms) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2697]: You can define it as PUT for creating "/sharedKeys/{sharedKeyName}" resource. Why are using Post API here?

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (visingla-ms) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2841]: [ARMBlockingComment]
The 202 response can't have schema.

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (visingla-ms) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:7976]: >DPD [](http://example.com/codeflow?start=26&length=3)

It will be good to put full form of this in description.

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (visingla-ms) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:9950]: This looks like a secret value, if yes, please use [https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-secret) here.

**Review by shengliumsft — APPROVED** [review]
Look good to me

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (shengliumsft) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2779]: > > sharedKey/reset
> 
> [ARMBlockingComment] This is not right way to define a POST Action. Either it should be "/vpnLinkConnections/{linkConnectionName}/resetSharedKey" or "/vpnLinkConnections/{linkConnectionName}/sharedKeys/{sharedKeyName}/reset".

We are following the API spec for Virtual Network Gateway Reset Shared Key for consistency. Please refer to the below link. It uses POST Action and "/connections/{virtualNetworkGatewayConnectionName}/sharedkey/reset".
[https://learn.microsoft.com/en-us/rest/api/network-gateway/virtual-network-gateway-connections/reset-shared-key?view=rest-network-gateway-2023-09-01&tabs=HTTP](url)
Example: POST https://management.azure.com/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/connections/{virtualNetworkGatewayConnectionName}/sharedkey/reset?api-version=2023-09-01

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (shengliumsft) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2637]: We are following the API spec for Virtual Network Gateway Get Shared Key for consistency. Please refer to the below link. 

https://learn.microsoft.com/en-us/rest/api/network-gateway/virtual-network-gateway-connections/get-shared-key?view=rest-network-gateway-2023-09-01&tabs=HTTP

Example: GET https://management.azure.com/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/connections/{virtualNetworkGatewayConnectionName}/sharedkey?api-version=2023-09-01

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (shengliumsft) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2697]: We are following the API spec for Virtual Network Gateway Set Shared Key for consistency. Please refer to the below link

https://learn.microsoft.com/en-us/rest/api/network-gateway/virtual-network-gateway-connections/set-shared-key?view=rest-network-gateway-2023-09-01&tabs=HTTP

Example: PUT https://management.azure.com/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/connections/{virtualNetworkGatewayConnectionName}/sharedkey?api-version=2023-09-01

It uses the PUT and "/connections/{virtualNetworkGatewayConnectionName}/sharedkey/", but we got the comment from API spec reviewers that Set/Reset Key operation should both be implemented as POST action. So, we changed it to POST action. Please refer to this API spec document for more information.
https://microsoft-my.sharepoint.com/:w:/p/liusheng/ESV5acTQcilMhbdQM3ZiCnIBBdbfEnZwUK4XfxhIzXgWTg?e=dK3wa3&nav=eyJjIjo4NDIzNDQyMjl9

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (Khushboo-Baheti) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:9950]: Fixed it

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (Khushboo-Baheti) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json]: added full form in description

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (Khushboo-Baheti) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2841]: changed it to 201

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (mentat9) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2637]: Unfortunately, that API doesn't follow ARM guidelines. Since you are wrapping the other API, it's not clear that consistency with a non ARM-conforming API is better than consistency with other ARM resources. To design this correctly for ARM, you could model it as a singleton resource named sharedKeys/default. You would need to support singleton GET, PUT, and a reset POST action (all on the sharedKeys/default resource), plus collection GET on the sharedKeys resource type (would always return a collection of one item). Your backend would need to slightly reshape the requests to talk to the other interface, but that should be simple.

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (mentat9) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2841]: See above comment. 201 isn't a valid POST status code for ARM.

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (mentat9) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2748]: >responses [](http://example.com/codeflow?start=9&length=9)

201 is not a valid POST response. Please follow the RPC for async POST. Ref: [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#call-action-post-asynchronously](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#call-action-post-asynchronously). Also, POST async response can't have a payload.

Comment (Khushboo-Baheti) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2748]: Thanks for sharing this. I fixed it

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (Khushboo-Baheti) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2637]: @mentat9 Sharing doc where we got Hybrid API review signoff [GetVpnConnectionSharedKeyAPI.docx](https://microsoft-my.sharepoint.com/:w:/p/liusheng/ESV5acTQcilMhbdQM3ZiCnIBBdbfEnZwUK4XfxhIzXgWTg?e=zutH2D)
We have a case opened for this and customers are waiting for this. Please provide a solution here

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (mentat9) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2755]: >201 [](http://example.com/codeflow?start=11&length=3)

This one isn't fixed yet. For async, this should be 202 with no schema. The other one below looks fine.

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (mentat9) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2637]: I'm not familiar with the Hybrid API review process. Regarding ARM compliance: please consider this feedback in future API designs. Extending non-ARM compliant APIs from one RP to another is not something we want, even to maintain consistency between those two. We are trying to improve the overall Azure API surface area by raising the quality and consistency of new APIs as they are added. Not blocking on this, since you are passing through another RPs API.

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (mentat9) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2697]: This needs to be fixed. You can't put POST on a resource type. If you want to implement this as a POST action, it needs to be a verb and applied to a resource instance, not a [broken] resource type. You could use this path for this POST action:

/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/vpnGateways/{gatewayName}/vpnConnections/{connectionName}/vpnLinkConnections/{linkConnectionName}/resetSharedKey

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (Khushboo-Baheti) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json]: @mentat9  Fixed it Thanks

**Comment by mentat9** [comment]
@Khushboo-Baheti - I'm discussing this further with other API reviewers.

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (Khushboo-Baheti) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:2697]: Sheng already provided response. We will definitely follow this in future but can we get an exception for this time ?

File: specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json
Comment (rkmanda) [inline specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:9953]: If this is considered a secret, you must not return it in a get API. Instead you need to return it in response to a post action

Comment (rkmanda) [reply specification/network/resource-manager/Microsoft.Network/stable/2023-11-01/virtualWan.json:9953]: Also remove the ARMChangesRequested label after u respond

**Comment by GuptaVertika** [comment]
Please merge the PR by 03/29, you have already crossed the code freeze date.

**Pull request closed without merging**
